syntax = "proto2";

import "serialization/geometry.proto";
import "serialization/physics.proto";
import "serialization/quantities.proto";

package principia.serialization;

message CelestialJacobiKeplerian {
  required int32 celestial_index = 1;
  optional int32 parent_index = 2;
  optional KeplerianElements keplerian_elements = 3;
  required MassiveBody body = 4;
}

message FlightPlan {
  required Quantity initial_mass = 1;
  required Point initial_time = 2;
  required Pair initial_degrees_of_freedom = 12;
  required Point desired_final_time = 3;
  repeated Manoeuvre manoeuvre = 8;
  required Ephemeris.AdaptiveStepParameters adaptive_step_parameters = 11;

  // Pre-Cardano.
  reserved 4, 5, 6, 7;
  reserved "integrator",
           "length_integration_tolerance",
           "segment",
           "speed_integration_tolerance";
  // Pre-Буняковский.
  reserved 9, 10;
  reserved "anomalous_segments", "max_steps";
}

message Manoeuvre {
  required Quantity thrust = 1;
  required Quantity initial_mass = 2;
  required Quantity specific_impulse = 3;
  required Multivector direction = 4;
  required Quantity duration = 5;
  required Point initial_time = 6;
  required DynamicFrame frame = 7;
  // Required, addded in Чебышёв.
  optional bool is_inertially_fixed = 8 [default = true];
}

message Part {
  required fixed32 part_id = 4;
  required string name = 9;
  required Quantity mass = 2;
  required Multivector intrinsic_force = 5;
  optional int32 containing_pile_up = 6;
  optional Pair degrees_of_freedom = 1;
  required DiscreteTrajectory tail = 7;
  required bool tail_is_authoritative = 8;

  // Pre-Cardano.
  reserved 3;
  reserved "gravitational_acceleration_to_be_applied_by_ksp";
}

message PileUp {
  repeated fixed32 part_id = 1;
  required Quantity mass = 2;
  required Multivector intrinsic_force = 3;
  required DiscreteTrajectory psychohistory = 4;
  map<fixed32, Pair> actual_part_degrees_of_freedom = 5;
  map<fixed32, Pair> apparent_part_degrees_of_freedom = 6;
  // Added in Cartan.
  optional Ephemeris.AdaptiveStepParameters adaptive_step_parameters = 7;
  optional Ephemeris.FixedStepParameters fixed_step_parameters = 8;
}

message Plugin {
  message VesselAndProperties {
    required string guid = 1;
    required Vessel vessel = 2;
    required int32 parent_index = 3;
    required bool loaded = 5;
    required bool kept = 7;
    // Pre-Cardano.
    reserved 4;
    reserved "dirty";
  }
  message CelestialParenthood {
    required int32 index = 1;
    optional int32 parent_index = 2;
    // Added in Catalan.
    optional int32 ephemeris_index = 3;
  }
  repeated VesselAndProperties vessel = 1;
  map<fixed32, string> part_id_to_vessel = 16;
  repeated CelestialParenthood celestial = 10;
  required Ephemeris ephemeris = 7;
  required Ephemeris.FixedStepParameters history_parameters = 12;
  required Ephemeris.AdaptiveStepParameters prolongation_parameters = 13;
  required Ephemeris.AdaptiveStepParameters prediction_parameters = 14;
  required Quantity planetarium_rotation = 4;
  required Point game_epoch = 15;
  required Point current_time = 5;
  required int32 sun_index = 6;
  optional DynamicFrame pre_cauchy_plotting_frame = 11;
  repeated PileUp pile_up = 17;
  optional Renderer renderer = 18;  // Added in Cauchy.

  // Pre-Cardano.
  reserved 3;
  reserved "bubble";
  // Pre-Буняковский.
  reserved 8, 9;
  reserved "prolongation_integrator", "prediction_integrator";
  // Pre-Bourbaki.
  reserved 2;
  reserved "/*celestial*/";
}

// Added in Cauchy.
message Renderer {
  required DynamicFrame plotting_frame = 1;
}

message Vessel {
  required string guid = 13;
  required string name = 19;
  required MasslessBody body = 1;
  required Ephemeris.AdaptiveStepParameters 
      prediction_adaptive_step_parameters = 10;
  repeated Part parts = 14;
  repeated fixed32 kept_parts = 15;
  required DiscreteTrajectory psychohistory = 16;
  required bool psychohistory_is_authoritative = 17;
  required DiscreteTrajectory prediction = 18;
  optional FlightPlan flight_plan = 4;

  // Pre-Буняковский.
  reserved 2, 3, 5;
  reserved "/*prediction*/", "history_and_prolongation", "owned_prolongation";
  // Pre-Cardano.
  reserved 6, 7, 8, 9, 11, 12;
  reserved "history",
           "prolongation_adaptive_step_parameters",
           "history_fixed_step_parameters",
           "prediction_last_time",
           "is_dirty",
           "prediction_fork_time";
}
